lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -2562,128 +2565,113 @@ bool ReplaceOpenCLBuiltinPass::replaceHadd(Function &F, bool is_signed,
+        auto callee = F.getParent()->getOrInsertFunction(sclamp_name, clamp_ty);
+        auto clamp = builder.CreateCall(callee, {op, min, max});
+        return builder.CreateTrunc(clamp, ty);
       } else {
-        // Pseudo-code:
-        // c = a + b;
+        // Add:
+        // c = a + b
         // if (b < 0)
         //   c = c > a ? min : c;
         // else
-        //   c = c < a ? max : c;
+        //   c  = c < a ? max : c;
         //
-        unsigned bitwidth = ty->getScalarSizeInBits();
-        Constant *scalar_min = ConstantInt::get(
-            Call->getContext(), APInt::getSignedMinValue(bitwidth));
-        Constant *scalar_max = ConstantInt::get(
-            Call->getContext(), APInt::getSignedMaxValue(bitwidth));
-        Constant *min = scalar_min;
-        Constant *max = scalar_max;
+        // Sub:
+        // c = a - b;
+        // if (b < 0)
+        //   c = c < a ? max : c;
+        // else
+        //   c = c > a ? min : c;
+        Constant *min = ConstantInt::get(Call->getContext(),
+                                         APInt::getSignedMinValue(bitwidth));
+        Constant *max = ConstantInt::get(Call->getContext(),
+                                         APInt::getSignedMaxValue(bitwidth));
         if (auto vec_ty = dyn_cast<VectorType>(ty)) {
           min = ConstantVector::getSplat(vec_ty->getElementCount(), min);
           max = ConstantVector::getSplat(vec_ty->getElementCount(), max);
         }
-        auto zero = Constant::getNullValue(ty);
-        // Cannot add the nsw flag.
-        auto add = BinaryOperator::Create(Instruction::Add, op0, op1, "", Call);
-        auto add_gt_op0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SGT,
-                                          add, op0, "", Call);
-        auto min_clamp = SelectInst::Create(add_gt_op0, min, add, "", Call);
-        auto add_lt_op0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SLT,
-                                          add, op0, "", Call);
-        auto max_clamp = SelectInst::Create(add_lt_op0, max, add, "", Call);
-        auto op1_lt_0 = CmpInst::Create(Instruction::ICmp, CmpInst::ICMP_SLT,
-                                        op1, zero, "", Call);
-        result = SelectInst::Create(op1_lt_0, min_clamp, max_clamp, "", Call);
+        Value *op = nullptr;
+        if (is_add) {
+          op = builder.CreateAdd(a, b);
+        } else {
+          op = builder.CreateSub(a, b);
+        }
+        auto b_lt_0 = builder.CreateICmpSLT(b, Constant::getNullValue(ty));
+        auto op_gt_a = builder.CreateICmpSGT(op, a);
+        auto op_lt_a = builder.CreateICmpSLT(op, a);
+        auto neg_cmp = is_add ? op_gt_a : op_lt_a;
+        auto pos_cmp = is_add ? op_lt_a : op_gt_a;
+        auto neg_value = is_add ? min : max;
+        auto pos_value = is_add ? max : min;
+        auto neg_clamp = builder.CreateSelect(neg_cmp, neg_value, op);
+        auto pos_clamp = builder.CreateSelect(pos_cmp, pos_value, op);
```

Comment:
Maybe marginally simpler to just do the `is_add` ternary statement once to select between the two select instructions (since they end up being the same for add and sub)? Amounts to the same thing though, so feel free to ignore this.

```
        auto op_gt_clamp = builder.CreateSelect(op_gt_a, min, op);
        auto op_lt_clamp = builder.CreateSelect(op_lt_a, max, op);
        auto neg_clamp = is_add ? op_gt_clamp : op_lt_clamp;
        auto pos_clamp = is_add ? op_lt_clamp : op_gt_clamp;
```